Update ERC-7579: executeUserOp + multitype module onInstall/onUninstall (#661)
ethereum/ERCs: ERCS/erc-7579.md

function executeUserOp(PackedUserOperation calldata userOp, bytes32 userOpHash) external;
```

> If account chooses to implement `executeUserOp`, this method SHOULD execute `userOp.calldata` except 4 most significant bytes, which are reserved for `executeUserOp.selector` as per ERC-4337.

Reviewer: is this too vague? should we specify that it MUST be delegatecalled like accounts do today?

Author: Yes, I was thinking of this as well. 
I agree, that delegatecall'ing here is preferred, but I don't see why it is a MUST.

IMO, what is important for the standard, is that exactly calldata[0:4] is executed, so we're always sure which part of calldata we can parse for validation parameters. 
So in theory it doesn't matter how account executes , it is more important what it executes

Reviewer: in this case, it could eg be the case that calldata[4:24] is an address and calldata[24:] is calldata which is called; or calldata[4:36] is target and calldata[36:] is data; or any other way of encoding the data here; so this wouldnt really help much for validators right

Author: ok, I see your point. let me introduce changes

> If account chooses to implement `executeUserOp`, this method SHOULD ensure the account executes `userOp.calldata` except 4 most significant bytes, which are reserved for `executeUserOp.selector` as per ERC-4337. Thus the `userOp.callData[4:]` should represent a valid call to the account as per Contract ABI specification.  

Reviewer: ```suggestion
If an account chooses to implement `executeUserOp`, this method SHOULD ensure the account executes `userOp.calldata` except 4 most significant bytes, which are reserved for `executeUserOp.selector` as per ERC-4337. Thus the `userOp.callData[4:]` should represent the calldata for a valid call to the account. It is RECOMMENDED that the account executes a `delegatecall` in order to preserve the original `msg.sender` to the account. 
```

> Example:

Reviewer: ```suggestion

Example:
```

- MUST route to fallback handlers based on the function selector of the calldata
- MAY implement authorization control, which SHOULD be done via hooks 

> If the account is adding features via fallback, in context of requirements defined by this specification, it should be considered the same as if the account was implementing those features natively.

Reviewer: ```suggestion
If the account adds features via fallback, these should be considered the same as if the account was implementing those features natively.
```

Reviewer: can this lead to accounts not having these features if the right fallback isnt isntalled? is this a problem?

Author: I think yes, this can lead for the account to end up not having these features. It will make the account not compatible with this standard. I do not think this is a problem, however, maybe, it can be explicitly outlined above.

Something like 

> In this case, uninstalling such a fallback will lead for account to become incompatible with ERC-7579.

wdyt?

Author: I think yes, this can lead for the account to end up not having these features. It will make the account not compatible with this standard. I do not think this is a problem, however, maybe, it can be explicitly outlined above.

Something like 

> In this case, uninstalling such a fallback will lead for account to become incompatible with ERC-7579.

wdyt?

> ERC-165 support (see below) is one example of such an approach.

Reviewer: ```suggestion
ERC-165 support (see below) is one example of such an approach. Note, that it is only RECOMMENDED to implement view functions via fallback where this can lead to greater extensibility. It is NOT RECOMMENDED to implement core account logic via a fallback.
```

}
```

> Note: A single module that is of multiple types MAY decide to pass `moduleTypeId` inside `data` to `onInstall` and/or `onUninstall` methods, so those methods are able to properly handle installation/uninstallation for various types.

Reviewer: does this need to be in the standard? this seems like module specific behaviour that would be encoded offchain like installation/uninstallation data

Author: I don't think it SHOULD be here, but it MAY be here :)
I just found it to be useful for people implementing modules according to the standard. More of best practice, which, btw, is not only about encoding data, but about decoding it in the onInstall/onUninstall as well

}
```

> Note: A single module that is of multiple types MAY decide to pass `moduleTypeId` inside `data` to `onInstall` and/or `onUninstall` methods, so those methods are able to properly handle installation/uninstallation for various types.

Reviewer: might be worth to give an example here as its not 100% clear to me

you're thinking smth like this right:

```solidity
// Module.sol

function onInstall(bytes memory data) external {
    (uint256 moduleTypeId, bytes memory otherData) = abi.decode(data, (uint256,bytes));
}
```

Author: yes.
```
if(moduleTypeId==VALIDATOR) {
   // procedure for validator
} elseif(moduleTypeId==EXECUTOR) {
  // procedure for executor
}
```
